src/cgutils.cpp

```diff
@@ -2152,23 +2164,24 @@ static void emit_unionmove(jl_codectx_t &ctx, Value *dest, const jl_cgval_t &src
                 emit_unbox(ctx, store_ty, src, typ, dest, isVolatile);
             }
             else {
-                Value *src_ptr = data_pointer(ctx, src, T_pint8);
-                if (dest->getType() != T_pint8)
-                    dest = emit_bitcast(ctx, dest, T_pint8);
-                if (skip) // copy dest -> dest to simulate an undef value / conditional copy
-                    src_ptr = ctx.builder.CreateSelect(skip, dest, src_ptr);
+                Value *src_ptr = data_pointer(ctx, src);
+                unsigned nb = jl_datatype_size(typ);
                 unsigned alignment = julia_alignment(typ, 0);
-                emit_memcpy(ctx, dest, src_ptr, jl_datatype_size(typ), alignment, isVolatile, tbaa);
+                Value *nbytes = ConstantInt::get(T_size, nb);
+                if (skip) // copy dest -> dest to simulate an undef value / conditional copy
```

Comment:
It appears that LLVM recently hit this problem too, and are working on a real fix (especially due the performance problems this change may cause) https://reviews.llvm.org/D86815